homeassistant/components/switcher_kis/switch.py

```diff
@@ -0,0 +1,179 @@
+"""Home Assistant Switcher Component Switch platform."""
+
+from logging import getLogger
+from typing import Any, Callable, cast, Dict, Optional
+
+from homeassistant.components.switch import ATTR_CURRENT_POWER_W, SwitchDevice
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.exceptions import PlatformNotReady
+
+from . import (
+    ATTR_AUTO_OFF_SET, ATTR_DEVICE_NAME, ATTR_ELECTRIC_CURRNET,
+    ATTR_REMAINING_TIME, DATA_DEVICE, DOMAIN, SIGNAL_SWITCHER_DEVICE_UPDATE)
+
+_LOGGER = getLogger(__name__)
+
+DEPENDENCIES = ['switcher_kis']
+
+PROPERTIES_TO_DEVICE_ATTRIBUTES = {
+    'current_power_w': ATTR_CURRENT_POWER_W,
+    'electric_current': ATTR_ELECTRIC_CURRNET,
+    'remaining_time': ATTR_REMAINING_TIME,
+    'auto_off_set': ATTR_AUTO_OFF_SET,
+    'device_name': ATTR_DEVICE_NAME
+}
+
+
+async def async_setup_platform(hass: HomeAssistantType, config: Dict,
+                               async_add_entities: Callable,
+                               discovery_info: Dict) -> None:
+    """Set up the switcher platform for the switch component."""
+    if DOMAIN not in hass.data:
+        raise PlatformNotReady("No configuration data found.")
+
+    async_add_entities([SwitcherControl(hass.data[DOMAIN][DATA_DEVICE])])
+
+
+class SwitcherControl(SwitchDevice):
+    """Home Assistant switch entity."""
+
+    def __init__(self, device_data: Any) -> None:
+        """Initialize the entity."""
+        # pylint: disable=unused-import
+        from aioswitcher.devices import SwitcherV2Device
+        # pylint: enable=unused-import
+        self._self_initiated = False
+        self._device_data = device_data  # type: SwitcherV2Device
+        self._state = device_data.state
+        self._unsub_dispatcher = None  # type: Optional[Callable[..., Any]]
+
+    @property
+    def electric_current(self) -> float:
```

Comment:
I think these few property is over designed, the only usage I found is in `device_state_attributes`. Do we really need them as a property?

Reply:
@awarecan 
Actually that was the purpose of this property.
Without it the `for loop` iterating over the `PROPERTIES_TO_DEVICE_ATTRIBUTES` dictionary inside the `device_state_attributes` property is pointless.

I can of course cancel the constant dictionary and the for loop and do everything manually within `device_state_attributes` or just manually add the `electric_current` part manually.
Please advice on what would you recommend?
Thank you.

Reply:
I think it's fine.  Other platforms follow a similar pattern when there's lots of things to add to `state_attributes` and avoids the repetitive code of _if attribute not None then add_ statements.

Reply:
If it is only limit to use inside `device_state_attributes`, why we need those properties? Will my code have same function?

```diff
    @property
    def device_state_attributes(self) -> Dict:
        """Return the optional state attributes."""
        from aioswitcher.consts import WAITING_TEXT

        attribs = {}

        for prop, attr in PROPERTIES_TO_DEVICE_ATTRIBUTES.items():
-            value = getattr(self, prop)
+            value = getattr(self._device_data, prop)
            if value and value is not WAITING_TEXT:
                attribs[attr] = value

        return attribs
```

Reply:
This will definitely save me some annoying `mypy` warnings.
But I think that will make the other 3 properties redundant as-well.
```python
PROPERTIES_TO_DEVICE_ATTRIBUTES = {
    'current_power_w': ATTR_CURRENT_POWER_W,
    'electric_current': ATTR_ELECTRIC_CURRNET,
    'remaining_time': ATTR_REMAINING_TIME,
    'auto_off_set': ATTR_AUTO_OFF_SET
}
```
The `current_power_w` is of course inherited from the `homeassistant.components.switch.SwitchDevice` class.

Other then that, I'm not sure it'll make the code more readable, you'll have to know the `aioswitcher.devices.SwitcherV2Device` class to know the attributes data types.
Although the names are pretty much imply their contents...

Anyway, I can go either way.
Which ever gets this component accepted ;-).

Reply:
It is unworthy using complex design only to satisfy mypy. Feel free to use `# type: ignore`

Reply:
Fixed in this [commit](https://github.com/home-assistant/home-assistant/pull/22325/commits/02704fb2efa3d44e01d0ac39f7cf7a10cbc23cdc).
:-)